### ISSUE TITLE
Supporting an External Vulkan library (Contains example!)

### ISSUE THREAD

**Issue by Mervill** [opentk/opentk#412]
Hi!

I've been working on a C# Vulkan library called [Tanagra](https://github.com/Mervill/Tanagra), and I've been using some of the `OpenTK.Math` objects (`Vector` and `Matrix`) to test it. I'd encourage you guys to check it out!

Now I'm starting to look at cross platform issues like opening a window, dealing with resize events and processing input. OpenTK is in the perfect position to support Vulkan without introducing massive changes to existing code.

In fact, one of the first things OpenTK would need to do to support Vulkan is remove a bunch of it's OpenGL dependencies. There's already been some discussion about splitting OpenTK into smaller libraries (seealso https://github.com/opentk/opentk/issues/360), and I'd recommend that `OpenTK.Math`, `OpenTK.Audio` and `OpenTK.Compute` (among others) be split into separate packages or otherwise removed from the main OpenTK dll (seealso https://github.com/opentk/opentk/pull/411).

Vulkan departs from OpenGL in a number of ways. Vulkan Instances need not be created on the main thread, and Vulkan commands can be called from any thread with only a [few restrictions](https://www.khronos.org/registry/vulkan/specs/1.0/xhtml/vkspec.html#fundamentals-threadingbehavior) on synchronization that are already dealt with (or plan on being dealt with) by my library. Vulkan is also very 'self-sufficient', meaning that it requires very little information from the host system that it can't discover on it's own. Whereas in OpenGL there is a lot of work that needs to be done up front to establish a valid Context, Vulkan can bootstrap itself and begin rendering with almost no outside information provided to it.

To render to a window (or any other surface) Vulkan simply needs to be provided a valid `VkSurfaceKHR` object appropriate for the window manager running on the system. This is done by calling a function like `vkCreate*SurfaceKHR` where the `*` is (currently) one of: `Android`, `DisplayPlane`, `Mir`, `Wayland`, `Win32`, `Xlib` or `Xcb`. The immediate implication of this is that the _driver vendor_ not the implementer, is given the responsibility of ensuring that rendered frames are correctly presented to the target window (resize and other window events are still the responsibility of the implementer, however). Both the Intel and NVidia drivers currently only support `Win32` or `Xcb` on Linux, both have said they plan to support `Xlib` and `DisplayPlane` (rendering without a specific WM) in the future.

Each of the surface creation functions takes a `*SurfaceCreateInfoKHR` object that, with the exception of `DisplayPlane`, requires only the pointer to the WM instance and the pointer to the current window. For example `Win32SurfaceCreateInfoKHR` only needs the `HINSTANCE` and `HWND` pointers and `XlibSurfaceCreateInfoKHR` only needs the `Dpy` pointer returned by `XOpenDisplay` and the window pointer returned by `XCreateSimpleWindow` (or another X11 create window function).

**So at the moment, OpenTK only needs to implement a way to retrieve these pointers** from the underlying `NativeWindow` object in order to minimally support Vulkan. As development continues the initialization process may become more complex, but as a proof of concept / MVP things could hardly be simpler.

To that end I've throw together an extremely quick-and-dirty MVP of OpenTK running a small example rendered with my Vulkan library. [You can check it out here](https://github.com/Mervill/opentk) (Open `OpenTK.Vulkan.sln` and run the example). Win32 / Latest Vulkan drivers and the [LunarG Vulkan SDK](https://lunarg.com/vulkan-sdk/) are required.

(seealso https://github.com/opentk/opentk/issues/342)

**Comment by Mervill** [comment]
So moving forward, here's what I'd suggest:

- Window acquire / events / input should be it's own package, `OpenTK.Platform`, this package would ideally not care about weather OpenGL or Vulkan will be used.
- The core `OpenTK` package could have `#if VULKAN` added to it to speficy the Vulkan build target, removing any remaining OpenGL dependencies (that can't be easily split).
- `OpenTK.OpenCL`, `OpenTK.OpenAL` and `OpenTK.Math` split into their own packages
- An `OpenTK.Vulkan` package (or `#if VULKAN` additions to `OpenTK`) would contain platform-dependent ways to create the `VkSurfaceKHR` object

A word about .NET versions: I'd like Vulkan to ship as a `4.6.X` library. As long as we're smart about how we split the packages (and perhaps a sneaky use of `InternalsVisibleTo`) this shouldn't conflict with the `2.0` target of OpenTK itself.

**Comment by cra0zy** [comment]
I am a bit confused why you opened a new issue instead of commenting on #342 ... anyway closing since this is a duplicate issue.

**Comment by varon** [comment]
@cra0zy They're quite similar, but distinct. This issue deals with adjusting the library to allow for interaction with an external Vulkan library, while #342 involves adding support for Vulkan directly to OpenTK.

**Comment by cra0zy** [comment]
This is a sub set of #342, why would you want to keep a separate issue about it?

**Comment by varon** [comment]
From what I understand of reading the code, this would involve API changes not required for the other one. It also takes a different approach by deferring the Vulkan bindings to a separate library. The title has been updated to reflect this.

**Comment by varon** [comment]
@Mervill This has been sitting for a while!

I'm totally in favour of adding Vulkan bindings, but I think we'll need the following before we can add them:
 - We have a working .Net core build
 - We've reviewed the black magic IL rewrite phase
 - We've split up the library a bit, into something like OpenTK.Math, OpenTK.GL, OpenTK.GLES, etc.

With that split, adding OpenTK.Vulkan would be a fairly trivial addition.

**Comment by varon** [comment]
Closing. We would really like to add Vulkan bindings, We're just too short on manpower at the moment considering what needs to happen first for this.
